README.md
[![Build Status](https://travis-ci.org/dex4er/nss-docker.png?branch=master)](https://travis-ci.org/dex4er/nss-docker)

This is a plugin for the GNU Name Service Switch (NSS) functionality of
the GNU C Library (glibc) providing mechanism for finding Docker
containers by theirs IDs.

The container names are searched in virtual domain name ".docker".

Install it and add this library to your `/etc/nsswitch.conf` file:

```
hosts: files docker [NOTFOUND=return] dns
```

Then try:

```
$ docker run -d --name test dex4er/alpine-init
a548fba14c314e77a1630580f6536e09002d84e5610b85a08c622fa451a4f893

$ getent hosts test.docker
172.17.0.4      test.docker

$ ping test.docker
PING test.docker (172.17.0.4) 56(84) bytes of data.
64 bytes from 172.17.0.4: icmp_seq=1 ttl=64 time=0.171 ms
```

This module does not require any additional libraries beside GNU C Library.

## Installation

#### From source archive:

Check the number of the latest release: https://github.com/dex4er/nss-docker/releases/latest

Then (ie. for 0.01 release):

```
$ curl -L https://github.com/dex4er/nss-docker/archive/0.01.tar.gz | tar zxv
$ ./configure --prefix=/usr
$ make
$ sudo make install
```

#### From Git repository:

```
$ git clone https://github.com/dex4er/nss-docker
$ ./autogen.sh
$ ./configure --prefix=/usr
$ make
$ sudo make install
```

## Testing

The tests requires root priviliges (for real chroot because `/etc/nsswitch.conf` can't be changed to another file)

It is suggested to compile necessary files (ie.: mock server) as standard user and run tests as root user.

```
$ make -C test check-src
$ sudo make test
```


NEWS.md
## Version 0.02

Jun 29 2016

* This version supports `--net` option.
* API 1.21 for Docker 1.7 is required.

## Version 0.01

Oct 21 2015

* First release


test/testtree.sh
#!/bin/sh

if [ $# -gt 0 ]; then
    destdir=$1
    shift
else
    destdir=testtree
fi

for d in \
    /bin \
    /etc \
    /lib \
    /lib/*-*-* \
    /lib32 \
    /lib64 \
    /usr/lib \
    /usr/lib/*-*-* \
    /usr/lib32 \
    /usr/lib64 \
    /libexec \
    /root \
    /sbin \
    /tmp \
    /usr/bin \
    /usr/lib \
    /usr/sbin \
    /usr/local/bin \
    /usr/local/lib \
    /var/run
do
    test -d $d || continue
    mkdir -p $destdir/$d
done

for d in \
    /dev \
    /proc
do
    rm -f $destdir/$d
    ln -sf $d $destdir/$d
done

for p in \
    '/bin/bash' \
    '/bin/busybox' \
    '/bin/cat' \
    '/bin/chmod' \
    '/bin/csh' \
    '/bin/cp' \
    '/bin/dash' \
    '/bin/echo' \
    '/bin/grep' \
    '/bin/ln' \
    '/bin/ls' \
    '/bin/mkdir' \
    '/bin/ps' \
    '/bin/pwd' \
    '/bin/readlink' \
    '/bin/rm' \
    '/bin/sed' \
    '/bin/sh' \
    '/bin/sleep' \
    '/bin/touch' \
    '/usr/bin/basename' \
    '/usr/bin/dirname' \
    '/usr/bin/env' \
    '/usr/bin/find' \
    '/usr/bin/getent' \
    '/usr/bin/id' \
    '/usr/bin/ischroot' \
    '/usr/bin/less' \
    '/usr/bin/ltrace' \
    '/usr/bin/more' \
    '/usr/bin/perl' \
    '/usr/bin/readlink' \
    '/usr/bin/seq' \
    '/usr/bin/sort' \
    '/usr/bin/strace' \
    '/usr/bin/test' \
    '/usr/bin/touch' \
    '/usr/bin/tr' \
    '/usr/sbin/chroot' \
    '/usr/local/bin/bash' \
    '/usr/local/bin/gseq' \
    '/usr/local/bin/ltrace' \
    '/usr/local/bin/strace'
do
    for f in $p; do
        d=$(dirname $f)
        if [ ! -e $f ]; then
                f=$(command -v $(basename $f) 2> /dev/null )
                if [ $? -ne 0 ]; then
                        continue
                fi
        fi
        cp -pf $PREFIX$f $destdir/$d 2>/dev/null
    done
done

for p in \
    'ld-*.so' \
    'ld-*.so.*' \
    'ld-linux.so.*' \
    'ld-linux-x86-64.so.*' \
    'ld-uClibc.so.*' \
    'ld.so.*' \
    'libacl.so.*' \
    'libattr.so.*' \
    'libc.so.*' \
    'libcrypt.so.*' \
    'libdl.so.*' \
    'libedit.so.*' \
    'libelf.so.*' \
    'libgcc_s.so.*' \
    'libm.so.*' \
    'libncurses.so.*' \
    'libpcre.so.*' \
    'libpthread.so.*' \
    'librt.so.*' \
    'libselinux.so.*' \
    'libtinfo.so.*' \
    'libutil.so.*' \
    'linux-vdso.so.*'
do
    for d in \
        /lib \
        /lib/*-*-* \
        /lib32 \
        /lib64 \
        /usr/lib \
        /usr/lib/*-*-* \
        /usr/lib32 \
        /usr/lib64 \
        /usr/local/lib
    do
        test -d "$d" || continue
        for f in $d/$p; do
            cp -pf $PREFIX$f $destdir/$(dirname $f) 2>/dev/null
        done
    done
done

for p in \
    src/test-*
do
    test -x $p || continue
    cp -pf $p $destdir/bin
done

cp -pf ../src/.libs/*.so* $destdir/lib

chflags -R 0 $destdir 2>/dev/null || true

echo $destdir > $destdir/CHROOT


test/src/test-docker-api-client.c
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef SUN_LEN
#define SUN_LEN(su) (sizeof(*(su)) - sizeof((su)->sun_path) + strlen((su)->sun_path))
#endif

#define DOCKER_API_REQUEST "GET /v%s/containers/%s/json HTTP/1.0\015\012\015\012"


int main(int argc, char *argv[]) {
    int sockfd, servlen, n;
    struct sockaddr_un serv_addr;
    char buffer[10240];

    if (argc != 4) {
        fprintf(stderr, "Usage: %s docker_socket api_version container_id\n", argv[0]);
        exit(2);
    }

    memset((char *) &serv_addr, 0, sizeof(serv_addr));
    serv_addr.sun_family = AF_UNIX;
    strcpy(serv_addr.sun_path, argv[1]);
    servlen = SUN_LEN(&serv_addr);

    if ((sockfd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
        perror("socket");
        exit(1);
    }

    if (connect(sockfd, (struct sockaddr *) &serv_addr, servlen) < 0) {
        perror("connect");
        exit(1);
    }

    memset(buffer, 0, 10240);
    snprintf(buffer, 10240, DOCKER_API_REQUEST, argv[2], argv[3]);
    if (write(sockfd, buffer, strlen(buffer)) < 0) {
        perror("write(newsockfd)");
        exit(1);
    }

    n = read(sockfd, buffer, 10240);
    if (write(1, buffer, n) < 0) {
        perror("write(newsockfd)");
        exit(1);
    }

    return 0;
}


test/src/test-docker-api-server.c
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>

#ifndef SUN_LEN
#define SUN_LEN(su) (sizeof(*(su)) - sizeof((su)->sun_path) + strlen((su)->sun_path))
#endif

#define DOCKER_RESPONSE_SUCCESS "HTTP/1.0 200 OK\r\nContent-Type: application/json\r\nDate: Sun, 20 Sep 2015 19:53:31 GMT\r\n\r\n{\"Id\":\"12b2dd585ecbdf168ad8a3645744445afe58794af7fdf2ab2b1bb97b6af3fc4a\",\"Created\":\"2015-09-20T11:24:42.329765535Z\",\"Path\":\"/sbin/init\",\"Args\":[],\"State\":{\"Running\":true,\"Paused\":false,\"Restarting\":false,\"OOMKilled\":false,\"Dead\":false,\"Pid\":28145,\"ExitCode\":0,\"Error\":\"\",\"StartedAt\":\"2015-09-20T11:24:42.850759517Z\",\"FinishedAt\":\"0001-01-01T00:00:00Z\"},\"Image\":\"55c1102409b9642a20b5f1d1d894a8419aaaf539e867005d36ac4ffd59abed012a2\",\"NetworkSettings\":{\"Bridge\":\"\",\"EndpointID\":\"4fb52aafa1874afea14712e1a71ab85ffec7d508f5f5b2b302bbf54a1e398853\",\"Gateway\":\"172.17.42.1\",\"GlobalIPv6Address\":\"\",\"GlobalIPv6PrefixLen\":0,\"HairpinMode\":false,\"IPAddress\":\"%s\",\"IPPrefixLen\":16,\"IPv6Gateway\":\"\",\"LinkLocalIPv6Address\":\"\",\"LinkLocalIPv6PrefixLen\":0,\"MacAddress\":\"02:42:ac:11:00:02\",\"NetworkID\":\"ea191dc39711f748a4b26d03abdb98817135b65a53d3c2beba6586d9ffceaf31\",\"PortMapping\":null,\"Ports\":{},\"SandboxKey\":\"/var/run/docker/netns/12b2dd585ecb\",\"SecondaryIPAddresses\":null,\"SecondaryIPv6Addresses\":null},\"ResolvConfPath\":\"/var/lib/docker/containers/12b2dd585ecbdf168ad8a3645744445afe58794af7fdf2ab2b1bb97b6af3fc4a/resolv.conf\",\"HostnamePath\":\"/var/lib/docker/containers/12b2dd585ecbdf168ad8a3645744445afe58794af7fdf2ab2b1bb97b6af3fc4a/hostname\",\"HostsPath\":\"/var/lib/docker/containers/12b2dd585ecbdf168ad8a3645744445afe58794af7fdf2ab2b1bb97b6af3fc4a/hosts\",\"LogPath\":\"/var/lib/docker/containers/12b2dd585ecbdf168ad8a3645744445afe58794af7fdf2ab2b1bb97b6af3fc4a/12b2dd585ecbdf168ad8a3645744445afe58794af7fdf2ab2b1bb97b6af3fc4a-json.log\",\"Name\":\"/%s\",\"RestartCount\":0,\"Driver\":\"aufs\",\"ExecDriver\":\"native-0.2\",\"MountLabel\":\"\",\"ProcessLabel\":\"\",\"Volumes\":{},\"VolumesRW\":{},\"AppArmorProfile\":\"\",\"ExecIDs\":null,\"HostConfig\":{\"Binds\":null,\"ContainerIDFile\":\"\",\"LxcConf\":[],\"Memory\":0,\"MemorySwap\":0,\"CpuShares\":0,\"CpuPeriod\":0,\"CpusetCpus\":\"\",\"CpusetMems\":\"\",\"CpuQuota\":0,\"BlkioWeight\":0,\"OomKillDisable\":false,\"Privileged\":false,\"PortBindings\":{},\"Links\":null,\"PublishAllPorts\":false,\"Dns\":null,\"DnsSearch\":null,\"ExtraHosts\":null,\"VolumesFrom\":null,\"Devices\":[],\"NetworkMode\":\"bridge\",\"IpcMode\":\"\",\"PidMode\":\"\",\"UTSMode\":\"\",\"CapAdd\":null,\"CapDrop\":null,\"RestartPolicy\":{\"Name\":\"no\",\"MaximumRetryCount\":0},\"SecurityOpt\":null,\"ReadonlyRootfs\":false,\"Ulimits\":null,\"LogConfig\":{\"Type\":\"json-file\",\"Config\":{}},\"CgroupParent\":\"\"},\"Config\":{\"Hostname\":\"12b2dd585ecb\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":true,\"AttachStderr\":true,\"PortSpecs\":null,\"ExposedPorts\":null,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"],\"Cmd\":null,\"Image\":\"hello-world\",\"Volumes\":null,\"VolumeDriver\":\"\",\"WorkingDir\":\"\",\"Entrypoint\":[\"/hello\"],\"NetworkDisabled\":false,\"MacAddress\":\"\",\"OnBuild\":null,\"Labels\":{},\"Memory\":0,\"MemorySwap\":0,\"CpuShares\":0,\"Cpuset\":\"\"}}\n"
#define DOCKER_RESPONSE_NOTFOUND "HTTP/1.0 404 Not Found\r\nContent-Type: text/plain; charset=utf-8\r\nDate: Sun, 20 Sep 2015 19:58:43 GMT\r\n\r\nno such id: %s\n"


int main(int argc, char *argv[]) {
    int sockfd, newsockfd, servlen, n, i, how_many;
    socklen_t clilen;
    struct sockaddr_un cli_addr, serv_addr, name_addr;
    socklen_t namelen;
    char buffer[10240];
    char *begin_container, *end_container;
    size_t containerlen;

    if (argc != 4 && argc != 5) {
        fprintf(stderr, "Usage: %s how_many docker_socket container_id ip_address\n", argv[0]);
        exit(2);
    }

    how_many = atoi(argv[1]);

    if ((sockfd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
        perror("socket");
        exit(1);
    }

    memset((char *) &serv_addr, 0, sizeof(serv_addr));
    serv_addr.sun_family = AF_UNIX;
    strcpy(serv_addr.sun_path, argv[2]);
    servlen = SUN_LEN(&serv_addr);
    if (bind(sockfd, (struct sockaddr *)&serv_addr, servlen) < 0) {
        perror("bind");
        exit(1);
    }

    namelen = SUN_LEN(&name_addr);
    if (getsockname(sockfd, (struct sockaddr*)&name_addr, &namelen) < 0) {
        perror("getsockname");
        exit(1);
    }

    listen(sockfd, 1);
    clilen = sizeof(cli_addr);

    for (i = 0; how_many == 0 || i < how_many; i++) {
        newsockfd = accept(sockfd, (struct sockaddr *) &cli_addr, &clilen);
        if (newsockfd < 0) {
            perror("accept");
            exit(1);
        }

        n = read(newsockfd, buffer, 10240);
        if (write(1, buffer, n) < 0) {
            perror("write(1)");
            exit(1);
        }

        if ((begin_container = strstr(buffer, "/containers/")) == NULL) {
            exit(1);
        }

        begin_container += sizeof("/containers/") - 1;

        if ((end_container = strchr(begin_container, '/')) == NULL) {
            exit(1);
        }

        *end_container = '\0';

        if ((containerlen = end_container - begin_container) == 0) {
            sprintf(buffer, DOCKER_RESPONSE_NOTFOUND, "CONTAINER");
        } else if (strncmp(begin_container, argv[3], containerlen) != 0) {
            sprintf(buffer, DOCKER_RESPONSE_NOTFOUND, begin_container);
        } else {
            sprintf(buffer, DOCKER_RESPONSE_SUCCESS, argv[4], argv[3]);
        }

        if (write(newsockfd, buffer, strlen(buffer)) < 0) {
            perror("write(newsockfd)");
            exit(1);
        }

        close(newsockfd);
    }

    return 0;
}


src/libnss_docker.c
/*
 * nss-docker - NSS plugin for looking up Docker containers
 *
 * Copyright (c) 2015-2016 Piotr Roszatycki <dexter@debian.org>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
 */

#include <config.h>

#define _BSD_SOURCE
#define _DEFAULT_SOURCE
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <assert.h>
#include <limits.h>
#include <nss.h>
#include <netdb.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#ifndef DEBUG
#define DEBUG 0
#endif

/* defined in config.h */
/* #define DOCKER_SOCKET "/var/run/docker.sock" */
/* #define DOCKER_API_VERSION "1.21" */
/* #define DOCKER_DOMAIN_SUFFIX ".docker" */

#ifndef HOST_NAME_MAX
#define HOST_NAME_MAX 255
#endif

#define ALIGN(a) (((a+sizeof(void*)-1)/sizeof(void*))*sizeof(void*))

#ifndef SUN_LEN
#define SUN_LEN(su) (sizeof(*(su)) - sizeof((su)->sun_path) + strlen((su)->sun_path))
#endif

#define STR_HELPER(x) #x
#define STR(x) STR_HELPER(x)

#define DOCKER_API_REQUEST "GET /v" DOCKER_API_VERSION "/containers/%." STR(HOST_NAME_MAX) "s/json HTTP/1.0\015\012\015\012"
#define HTTP_404 "HTTP/1.0 404"
#define FIND_NETWORKS ",\"Networks\":{"
#define FIND_IPADDRESS ",\"IPAddress\":\""


enum nss_status _nss_docker_gethostbyname3_r(
    const char *name, int af,
    struct hostent *result, char *buffer, size_t buflen, int *errnop,
    int *herrnop, int32_t *ttlp, char **canonp
) {
    /* full name length */
    size_t name_len;

    /* hostname is name without domain suffix */
    char hostname[256];

    /* docker domain suffix starts here */
    char *hostname_suffix_ptr;

    /* address of Docker API server */
    struct sockaddr_un docker_api_addr;

    /* length of docker_api_addr structure */
    socklen_t docker_api_addr_len;

    /* calculated buffer size */
    size_t buffer_size;

    /* offset for buffer ptr */
    size_t buffer_offset;

    /* socket descriptor */
    int sockfd;

    /* request message buffer */
    char req_message_buffer[sizeof(DOCKER_API_REQUEST) + HOST_NAME_MAX];

    /* request message size */
    size_t req_message_len;

    /* response message buffer */
    char res_message_buffer[10240];

    /* response message size */
    size_t res_message_len;

    /* Buffer for IPAddress string value */
    char ipaddress_str[16];

    /* Pointer for begin of Networks value */
    char *begin_networks;

    /* Pointers for begin and end of IPAddress value */
    char *begin_ipaddress, *end_ipaddress;

    /* Length of IPAddress value */
    size_t ipaddress_len;

    /* IPAddress as in_addr */
    struct in_addr ipaddress_addr;

    /* List of aliases in hostent */
    char *aliases;

    /* First address in hostent */
    char *addr_ptr;

    /* List of addresses in hostent */
    char *addr_list;

    if (DEBUG) fprintf(stderr, "_nss_docker_gethostbyname3_r(name=\"%s\", af=%d)\n", name, af);

    /* Handle only IPv4 */
    if (af != AF_INET) {
        *errnop = EAFNOSUPPORT;
        goto return_unavail;
    }

    /* Basic assertions for host name */
    name_len = strlen(name);

    if (name_len == 0) {
        *errnop = EADDRNOTAVAIL;
        goto return_unavail;
    }

    if (name_len > 255) {
        name_len = 255;
    }

    strncpy(hostname, name, sizeof(hostname));
    hostname[name_len] = '\0';

    /* Handle only .docker domain */
    if ((hostname_suffix_ptr = strstr(hostname, DOCKER_DOMAIN_SUFFIX)) == NULL) {
        *errnop = EADDRNOTAVAIL;
        goto return_unavail;
    }

    if (hostname_suffix_ptr[sizeof(DOCKER_DOMAIN_SUFFIX) - 1] != '\0') {
        *errnop = EADDRNOTAVAIL;
        goto return_unavail;
    }

    *hostname_suffix_ptr = '\0';

    /* Connect to Docker API socket */
    memset((char *) &docker_api_addr, 0, sizeof(docker_api_addr));
    docker_api_addr.sun_family = AF_UNIX;
    strcpy(docker_api_addr.sun_path, DOCKER_SOCKET);
    docker_api_addr_len = SUN_LEN(&docker_api_addr);

    if ((sockfd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
        *errnop = errno;
        goto return_unavail;
    }

    if (connect(sockfd, (struct sockaddr *) &docker_api_addr, docker_api_addr_len) < 0) {
        *errnop = errno;
        goto return_unavail;
    }

    /* Prepare request message */
    req_message_len = snprintf(req_message_buffer, sizeof(req_message_buffer) - 1, DOCKER_API_REQUEST, hostname);

    if (req_message_len == sizeof(req_message_buffer) - 1) {
        req_message_buffer[req_message_len] = '\0';
    }

    /* Send request */
    if (write(sockfd, req_message_buffer, strlen(req_message_buffer)) < 0) {
        close(sockfd);
        *errnop = errno;
        goto return_unavail;
    }

    /* Receive response */
    res_message_len = read(sockfd, res_message_buffer, sizeof(res_message_buffer) - 1);

    close(sockfd);

    if (res_message_len <= 0) {
        *errnop = errno;
        goto return_unavail;
    }

    res_message_buffer[res_message_len] = '\0';

    if (DEBUG) fwrite(res_message_buffer, res_message_len, 1, stderr);

    /* Handle HTTP 404 Not Found */
    if (strncmp(res_message_buffer, HTTP_404, sizeof(HTTP_404) - 1) == 0) {
        goto return_notfound;
    }

    /* Check if there is Networks key */
    if ((begin_networks = strstr(res_message_buffer, FIND_NETWORKS)) == NULL) {
        /* If not, will search from the beginning */
        begin_networks = res_message_buffer;
    }

    /* Check if there is IPAddress key */
    if ((begin_ipaddress = strstr(begin_networks, FIND_IPADDRESS)) == NULL) {
        *errnop = EBADMSG;
        goto return_unavail;
    }

    /* Check if it looks like IPAddress value */
    begin_ipaddress += sizeof(FIND_IPADDRESS) - 1;

    if (*begin_ipaddress == '"') {
        goto return_notfound;
    }

    if ((end_ipaddress = strchr(begin_ipaddress, '"')) == NULL) {
        *errnop = EBADMSG;
        goto return_unavail;
    }

    if ((ipaddress_len = end_ipaddress - begin_ipaddress) > 15) {
        *errnop = EBADMSG;
        goto return_unavail;
    }

    if (ipaddress_len == 0) {
        goto return_notfound;
    }

    strncpy(ipaddress_str, begin_ipaddress, ipaddress_len);
    ipaddress_str[ipaddress_len] = '\0';

    if (DEBUG) fprintf(stderr, "ipaddress_str=\"%s\"\n", ipaddress_str);

    /* Convert string to in_addr */
    if (! inet_aton(ipaddress_str, &ipaddress_addr)) {
        *errnop = EBADMSG;
        goto return_unavail;
    }

    /* Prepare hostent result */
    result->h_name = buffer;

    buffer_size = ALIGN(strlen(name) + 1) + sizeof(char*) + ALIGN(sizeof(struct in_addr));

    if (buflen < buffer_size) {
        *errnop = ENOMEM;
        goto return_tryagain;
    }

    strcpy(result->h_name, name);

    buffer_offset = ALIGN(strlen(name) + 1);

    aliases = buffer + buffer_offset;
    *(char **) aliases = NULL;
    buffer_offset += sizeof(char*);

    result->h_aliases = (char **) aliases;

    result->h_addrtype = AF_INET;
    result->h_length = sizeof(struct in_addr);

    addr_ptr = buffer + buffer_offset;
    memcpy(addr_ptr, &ipaddress_addr, result->h_length);
    buffer_offset += ALIGN(result->h_length);
    assert(buffer_offset == buffer_size);

    addr_list = buffer + buffer_offset;
    ((char **) addr_list)[0] = addr_ptr;
    ((char **) addr_list)[1] = NULL;

    result->h_addr_list = (char **) addr_list;

    return NSS_STATUS_SUCCESS;

return_unavail:
    *herrnop = NO_DATA;
    return NSS_STATUS_UNAVAIL;

return_tryagain:
    *herrnop = NO_RECOVERY;
    return NSS_STATUS_TRYAGAIN;

return_notfound:
    *errnop = ENOENT;
    *herrnop = HOST_NOT_FOUND;
    return NSS_STATUS_NOTFOUND;
}


enum nss_status _nss_docker_gethostbyname2_r(
    const char *name, int af,
    struct hostent *result, char *buffer, size_t buflen, int *errnop,
    int *herrnop
) {
    return _nss_docker_gethostbyname3_r(
        name, af, result, buffer, buflen,
        errnop, herrnop, NULL, NULL
    );
}


enum nss_status _nss_docker_gethostbyname_r(
    const char *name,
    struct hostent *result, char *buffer, size_t buflen, int *errnop,
    int *herrnop
) {
    int af = AF_UNSPEC;

    return _nss_docker_gethostbyname3_r(
        name, af, result, buffer, buflen,
        errnop, herrnop, NULL, NULL
    );
}
